File: server/models/product.js
const mongoos = require('mongoose');
const ratingSchema = require('./rating');

const productSchema = new mongoos.Schema({
    name: {
        type: String,
        trim: true,
        required: true
    },
    description: {
        type: String,
        trim: true,
        required: true
    },
    price: {
        type: Number,
        required: true
    },
    quantity: {
        type: Number,
        required : true
    },
    images: [
        {
            type: String,
            required: true
        }
    ],
    category: {
        type: String,
        required: true
    },
    ratings: [ratingSchema],    

});

const Product = new mongoos.model("Product", productSchema);

module.exports = {Product, productSchema};

File: lib/models/product.dart
import "dart:convert";

import "package:amazon_flutter/models/rating.dart";

class Product {
  final String name;
  final String description;
  final double quantity;
  final double price;
  final String category;
  final List<String> images;
  final String? id;
  final List<Rating>? rating;

  Product(
      {required this.name,
      required this.description,
      required this.quantity,
      required this.price,
      required this.category,
      required this.images,
      this.id,
      this.rating});

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'description': description,
      'quantity': quantity,
      'price': price,
      'category': category,
      'images': images,
      'id': id,
      'rating': rating
    };
  }

  factory Product.fromMap(Map<String, dynamic> map) {
    return Product(
        name: map['name'],
        description: map['description'],
        quantity: map['quantity'].toDouble() ?? 0.0,
        price: map['price'].toDouble() ?? 0.0,
        category: map['category'],
        images: List<String>.from(map['images']),
        id: map['_id'],
        rating: map['ratings'] != null
            ? List<Rating>.from(map['ratings']?.map((x) => Rating.fromMap(x)))
            : null);
  }

  String toJson() => jsonEncode(toMap());

  factory Product.fromjson(String source) =>
      Product.fromMap(jsonDecode(source));
}
